### PR TITLE
Add `BaseAPIClient`

### PR DESCRIPTION
**Pull request by pragmader**
Can be used to easily perform HTTP requests to various JSON
APIs. Also, it's going to be used in the code generator.

### CODE REVIEW

File: pkg/http/clients/base_api_client.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// APIError describes an error during unsuccessful HTTP request to an API
+type APIError struct {
+	// Status is the HTTP status of the API response
+	Status int
+	// Header is the set of response headers
+	Header http.Header
```
Comment (pragmader) [inline pkg/http/clients/base_api_client.go:26]: @LucasRoesler added the `Header` as promised.

File: pkg/http/clients/base_api_client.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+		// general error processing
+		response, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return errors.Wrap(err, "failed to read response body")
+		}
+		span.LogKV("response.body", string(response))
+		err = APIError{
+			Status:   resp.StatusCode,
+			Header:   resp.Header.Clone(),
```
Comment (pragmader) [inline pkg/http/clients/base_api_client.go:194]: This is how it's populated.

File: pkg/http/clients/base_api_client_test.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+			serverStatus:   http.StatusInternalServerError,
+			serverResponse: []byte("some crazy internal stuff"),
+
+			expError: APIError{
+				Status: http.StatusInternalServerError,
+				Header: http.Header{
+					"Content-Length": []string{"25"},
+					"Content-Type":   []string{"application/json"},
+					"Date":           []string{"fixed value"},
+				},
```
Comment (pragmader) [inline pkg/http/clients/base_api_client_test.go:185]: Here it's tested.

File: pkg/http/clients/base_api_client.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	// Response is the response body
+	Response []byte
+}
+
+// Error implements the error interface
+func (e APIError) Error() string {
+	return http.StatusText(e.Status)
+}
+
+// BaseAPIClient describes all basic HTTP client operations required to work with an API
```
Comment (LucasRoesler) [inline pkg/http/clients/base_api_client.go:36]:
```suggestion
// BaseAPIClient describes all basic HTTP client operations required to work with a JSON API
```

There are a couple spots where we are enforcing JSON req/resp, so we should be more specific here

**Review by LucasRoesler — CHANGES_REQUESTED** [review]

File: pkg/http/clients/base_api_client.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	if queryString != "" {
+		url += "?" + queryString
+	}
+
+	logrus := logrus.
+		WithField("method", method).
+		WithField("url", url)
+
+	logrus.Debug("creating the request token...")
+	token, err := t.tokenCreator.Create("tenjinManager", tokens.Options{Audience: "tenjin"})
```
[the remark this answers is not in the document]
Comment (pragmader) [reply pkg/http/clients/base_api_client.go:93]: good catch, forgot about it.

File: pkg/http/clients/base_api_client.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	url := t.GetBaseURL() + path
+	if queryString != "" {
+		url += "?" + queryString
+	}
+
+	logrus := logrus.
+		WithField("method", method).
+		WithField("url", url)
+
+	logrus.Debug("creating the request token...")
```
[the remark this answers is not in the document]
Comment (LucasRoesler) [reply pkg/http/clients/base_api_client.go:99]: this will still panic if the provider is nil

[the remark this answers is not in the document]
Comment (pragmader) [reply pkg/http/clients/base_api_client.go:99]: And it should, one must always use a provider, there is `NoopTokenProvider` if you need it.

[the remark this answers is not in the document]
Comment (LucasRoesler) [reply pkg/http/clients/base_api_client.go:99]: ok, i can't think of any specific counter arguments, let's keep the panic

**Review by LucasRoesler — DISMISSED** [review]

**Review by LucasRoesler — APPROVED** [review]

**Pull request merged**